src/Vector.cpp

```diff
@@ -0,0 +1,349 @@
+@return nullptr if the vector is empty.
+*/
+int *Vector::data() {
+    if (empty())
+        return nullptr;
+
+    return &m_data[0];
+}
+
+/**
+
+@brief Returns whether the vector is empty or not.
+@return true if the vector is empty, false otherwise.
+*/
+bool Vector::empty() const {
+    if (m_size == 0)
+        return true;
+
+    return false;
+}
+
+/**
+
+@brief Returns the size of the vector.
+@return size_t The size of the vector.
+*/
+size_t Vector::size() const {
+    return m_size;
+}
+
+/**
+
+@brief Returns the capacity of the vector.
+@return size_t The capacity of the vector.
+*/
+size_t Vector::capacity() const {
+    return m_capacity;
+}
+/**
+
+@brief Clears the vector and deallocates all memory associated with it.
+*/
+void Vector::clear() {
+    delete[] m_data;
+    m_size = 0;
+    m_capacity = 0;
+    m_data = new int [m_capacity];
+}
+/**
+
+@brief Resizes the vector to a new capacity and fills new elements with a given value.
+@param newCapacity The new capacity of the vector.
+@param value The value to be assigned to new elements.
+*/
+void Vector::resize(const size_t newCapacity, const int value) {
+    std::cout << "value is: " << value;
+    int* newData = new int[newCapacity];
+    if (newCapacity > m_size){
+        for (size_t i = 0; i < m_size; ++i)
+            newData[i] = m_data[i];
+
+        for (size_t i = m_size; i < newCapacity; ++i)
+            newData[i] = value;
+
+    }
+    else{
+        for (size_t i = 0; i < newCapacity; ++i)
+            newData[i] = m_data[i];
+    }
+
+    m_capacity = newCapacity;
+    delete[] m_data;
+    m_data = newData;
+    m_size = newCapacity;
+
+
+
+}
+/**
+
+@brief Swaps the contents of this vector with another vector.
+@param other The vector to swap with.
+*/
+void Vector::swap(Vector& other) {
+    // swap the internal data
+    int* temp_data = m_data;
+    m_data = other.m_data;
+    other.m_data = temp_data;
+
+    // swap the size
+    size_t temp_size = m_size;
+    m_size = other.m_size;
+    other.m_size = temp_size;
+
+    // swap the capacity
+    size_t temp_capacity = m_capacity;
+    m_capacity = other.m_capacity;
+    other.m_capacity = temp_capacity;
+}
+/**
+@brief Adds an element to the end of the vector.
+@param value The value of the element to add.
+*/
+void Vector::push_back(const int value) {
+    insert(m_size, value);
+}
+/**
+@brief Removes the last element from the vector.
+*/
+void Vector::pop_back() {
+    erase(m_size-1);
+}
+/* =================operators overloading=================*/
+/**
+
+@brief Overloaded operator[] for accessing elements in the vector by index.
+@param index The index of the element to access.
+@return int& A reference to the element at the specified index.
+@throws std::out_of_range if the index is out of range (i.e., less than 0 or greater than the current size of the vector).
+*/
+int &Vector::operator[](const size_t& index) {
+    if (index < 0 || index > m_size)
```

Comment:
`size_t` is an unsigned integer type, so no need to check if it's smaller than 0